apps/test/unit/code-studio/showProjectAdminTest.js

```diff
@@ -0,0 +1,119 @@
+import $ from 'jquery';
+import sinon from 'sinon';
+import {assert} from '../../util/configuredChai';
+import showProjectAdmin from '@cdo/apps/code-studio/showProjectAdmin';
+import {enforceDocumentBodyCleanup, replaceOnWindow, restoreOnWindow} from "../../util/testUtils";
+
+describe('showProjectAdmin', () => {
+  enforceDocumentBodyCleanup({checkEveryTest: true});
+
+  describe('' /* Wrapper describe allows testing cleanup properly */, () => {
```

Comment:
Ah, we should have cleaned this up.

The `enforceDocumentBodyCleanup({checkEveryTest: true});` helper adds a `beforeEach` and `afterEach` step in the `describe` scope that contains it.  If we then put beforeEach/afterEach steps that manipulate the DOM in the same scope:

```js
describe('outer scope', () => {
  enforceDocumentBodyCleanup({checkEveryTest: true});
  beforeEach(() => /* Add to DOM */);
  afterEach(() => /* Remove from DOM */);
});
```

...then we get failures even though our cleanup is good, because of the order in which the beforeEach/afterEach steps run:

```
enforceDocumentBodyCleanup records initial DOM
beforeEach adds test DOM
enforceDocumentBodyCleanup checks DOM against initial DOM, and fails
afterEach removes test DOM
```

By adding an inner describe scope we fix the ordering of these events:

```js
describe('outer scope', () => {
  enforceDocumentBodyCleanup({checkEveryTest: true});
  describe('inner scope', () => {
    beforeEach(() => /* Add to DOM */);
    afterEach(() => /* Remove from DOM */);
  });
});
```

```
enforceDocumentBodyCleanup records initial DOM
beforeEach adds test DOM
afterEach removes test DOM
enforceDocumentBodyCleanup checks DOM against initial DOM, and passes
```

I hadn't encountered this problem with this helper before, because [back when I introduced it](https://github.com/code-dot-org/code-dot-org/commit/afc52310c2e887ec8e3d3077190410c0dba9bd3a) my intention was to enable it in our top-level scope for all unit tests, not just for a small set of them.

Possible solutions here would be to remove this enforcement and the wrapper scope (it's most useful while working on the tests), or allowing the helper to take a callback and provide its own scope for everything in the callback.

Thoughts?

Reply:
Thanks for the great explanation! As for the options you listed, I'm inclined to go for the second one -> the helper is useful to have around and adding this extra functionality seems to make the helper even more useful. But for the expediency of this PR, I think just removing the enforcement works too.